### PR TITLE
Compare versions #31

### PR DESCRIPTION
**Pull request by AdamSpannbauer**
address issues in #31 

Summary of changes

New version comparison logic consists of 3 functions
  * `equivalent_version()` - Checks if wildcarded target version is equivalent to existing installed version (helper for `compare_version()`)
  * `compare_compatible_version()` - Check if existing installed version ~= target version (helper for `compare_version()`)
  * `compare_version()` - perform actual version comparisons; replaces `check_version()`
    * First 2 args swapped from what they were in `check_version()` (lines up with how the old `compare_version()` arguments were ordered).  All code referencing `check_version` has been updated to reference new `compare_version()`

### CODE REVIEW

**Comment by AdamSpannbauer** [comment]
This will have conflicts to be resolved with #38

File: R/requirements_helpers.R [hunk 1]
```diff
@@ -11,39 +11,97 @@ legal_r_package_name = function(name){
...
+  is_equality_check = comp %in% c('>=', '<=', '~=', '==')
+
+  # Check if `target` is equivalent to `existing` wrt wildcards
+  is_equivalent_version = equivalent_version(existing, target)
+
+  if (is_equality_check & is_equivalent_version) return(TRUE)
+  if (comp == "!=" & is_equivalent_version) return(FALSE)
+
+  # TODO: definitively prove this is a valid substitution
+  non_wild_card_target = gsub('*', '0', target, fixed = TRUE)
```
Comment (AdamSpannbauer) [inline R/requirements_helpers.R:96]: Want to ensure we have all possible edge cases covered in tests.  My brain is too fried rn to be able to think through if this substitution is safe at this point in the function.

File: R/requirements_helpers.R [hunk 2]
```diff
@@ -178,17 +236,17 @@ install_reqs = function(reqs, dryrun, verbose = dryrun,
...
-        available_compatibility = vapply(available_versions,
-                                         check_version,
-                                         TRUE,
-                                         target = version,
-                                         comp = comp)
+
+        available_compatibility = vapply(available_versions, function(available_version) {
+          compare_version(available_version, version, comp)
+        }, logical(1))
+
```
Comment (bmewing) [inline R/requirements_helpers.R:249]: Is there a reason why you're using an anonymous function instead of just passing in coompare_version with explicitly named arguments?

File: R/requirements_helpers.R (hunk 1 above)
Comment (bmewing) [reply R/requirements_helpers.R:96]: So if I understand this correctly, the requirement `mgsub <= 2.*` gets converted to `mgsub <= 2.0`, right?

File: R/requirements_helpers.R [hunk 3]
```diff
@@ -11,39 +11,97 @@ legal_r_package_name = function(name){
...
+  if (is_equality_check & is_equivalent_version) return(TRUE)
+  if (comp == "!=" & is_equivalent_version) return(FALSE)
+
+  # TODO: definitively prove this is a valid substitution
+  non_wild_card_target = gsub('*', '0', target, fixed = TRUE)
+
+  existing_version = package_version(existing)
+  target_version = package_version(non_wild_card_target)
+
```
Comment (bmewing) [inline R/requirements_helpers.R:100]: Nice find with the built-in package_version

File: R/requirements_helpers.R (hunk 1 above)
Comment (bmewing) [reply R/requirements_helpers.R:96]: The equivalent version check always starts at the front of the string, so you won't have an issue of `existing = '2.9'; target = '2.*'` mucking it up, so it seems like this should be sufficient. Given that all tests pass, it seems solid.

**Review by bmewing — APPROVED** [review]
Looks solid

File: R/requirements_helpers.R (hunk 2 above)
Comment (AdamSpannbauer) [reply R/requirements_helpers.R:249]: I personally find it more readable but I’m open to either way. Probably a hint slower to use the anonymous function.
